pyomo/repn/plugins/gams_writer.py

```diff
@@ -0,0 +1,512 @@
+        put_results = io_options.pop("put_results", None)
+
+        if len(io_options):
+            raise ValueError(
+                "ProblemWriter_gams passed unrecognized io_options:\n\t" +
+                "\n\t".join("%s = %s"
+                            % (k,v) for k,v in io_options.iteritems()))
+
+        if solver is not None:
+            if solver.upper() not in valid_solvers:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "solver: %s" % solver)
+
+        if mtype is not None:
+            valid_mtypes = set([
+                'lp', 'qcp', 'nlp', 'dnlp', 'rmip', 'mip', 'rmiqcp', 'rminlp',
+                'miqcp', 'minlp', 'rmpec', 'mpec', 'mcp', 'cns', 'emp'])
+            if mtype.lower() not in valid_mtypes:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "model type: %s" % mtype)
+            if (solver is not None and
+                mtype.upper() not in valid_solvers[solver.upper()]):
+                raise ValueError("ProblemWriter_gams passed solver (%s) "
+                                 "unsuitable for given model type (%s)"
+                                 % (solver, mtype))
+
+        if output_filename is None:
+            output_filename = model.name + ".gms"
+
+        if symbolic_solver_labels and (labeler is not None):
+            raise ValueError("ProblemWriter_gams: Using both the "
+                             "'symbolic_solver_labels' and 'labeler' "
+                             "I/O options is forbidden")
+
+        if symbolic_solver_labels:
+            var_labeler = con_labeler = AlphaNumericTextLabeler()
+        elif labeler is None:
+            var_labeler = NumericLabeler('x')
+            con_labeler = NumericLabeler('c')
+        else:
+            var_labeler = con_labeler = labeler
+
+        var_list = []
+        symbolMap = SymbolMap()
+
+        def var_recorder(obj):
+            ans = var_labeler(obj)
+            var_list.append(ans)
+            return ans
+
+        def var_label(obj):
+            if obj.is_fixed():
+                return str(value(obj))
+            return symbolMap.getSymbol(obj, var_recorder)
+
+        # when sorting, there are a non-trivial number of
+        # temporary objects created. these all yield
+        # non-circular references, so disable GC - the
+        # overhead is non-trivial, and because references
+        # are non-circular, everything will be collected
+        # immediately anyway.
+        with PauseGC() as pgc:
+            try:
+                if isinstance(output_filename, string_types):
+                    output_file = open(output_filename, "w")
+                else:
+                    # Support passing of stream such as a StringIO
+                    # on which to write the model file
+                    output_file = output_filename
+                if isinstance(model, IBlockStorage):
```

Comment:
> We are "saved" because CPython isn't really multithreaded 

I don't think this is correct, John. In CPython, the GIL just prevents most opportunities for CPU-bound threads to run at the same time. However, even in CPython, the GIL is released during blocking I/O (e.g., file.write), which is exactly what is happening in a problem writer. So this is not even safe in CPython.